### ISSUE TITLE
Unified type system for internal operations

### ISSUE THREAD

**Issue by re-xyr** [taichi-dev/taichi#5382]
Proposal for a unified type system and typechecking process for frontend AST expressions. This aims to solve:

- Inconsistent error messages;
- Overly relaxed typing rules;
- Code duplication in frontend AST.

This is a second attempt of simplifying internal operations after #4180; the type system described here is largely the same as that in #4180.

The proposal: [typing-taichi-internal-ops.pdf](https://github.com/taichi-dev/taichi/files/9080452/typing-taichi-internal-ops.pdf)

**Comment by re-xyr** [comment]
Here's a more informal introduction to the type system:

---

Soon, all internal operations in the frontend IR (including Unary/Binary/TernaryOpExpression, InternalFuncCallExpression etc) will all be merged into one expression: CallExpression, so that they are typechecked under a unified type system. This article provides a short informal introduction to that system.

## Giving operations types

Each operation (+, -, +=, ...) is given a type signature. For example, the type signature of `+` is:

```kt
<T : Primitive, U : Primitive> (lhs : T, rhs : U) -> T | U
```

This signature roughly means "accept two arguments that are of primitive types, and the return type is their common type". More precisely, it has 3 parts:

- The type variables list, and their traits: `<T : Primitive, U : Primitive>` means that we have two arbitrary types `T` and `U` that needs to be primitive (`i32`, `f32`, ...). This is similar to "generics" or "templates" in other languages.
- The parameter list: `(lhs : T, rhs : U)`, means that the operation accepts two arguments of the aforementioned primitive types.
- The return type: `-> T | U`. The type `T | U` means the common type of `T` and `U`, i.e. the type that is implicitly converted to.

## Type expressions

You probably have noticed that in the parameter list and return type, we specify the types not with simple data types like `i32`, but with something like an expression (`T` and `U` are variables, and `|` functions as an operator for types). This is a type expression. All kinds of type expression that can be used are listed below:

- A plain datatype like `i32` is a type expression;
- Two type expressions joined by the operator `|` means their common type;
- `comp(e)` where `e` is a type expression means its compute type;
- Type variables mentioned in the type variables list (like `T` and `U` in the previous example) can be used as a type expression.

With type expressions we can specify what types we want declaratively, which is much less error-prone than writing the typing rules out in code imperatively.

## Error messages

There are two kinds of possible error messages:

- Type mismatch: when the expected parameter type and the actual argument type are not the same.
  ```
  In a call to the operation `+`:
    expected i64 for argument 2, but got i32
    ( called with argument types i64, i32 )
  ```
- Trait not satisfied: when the argument type did not satisfy a trait that is required.
  ```
  In a call to the operation `+`:
    the argument type CustomInt(15) is not primitive
  ```
